### PR TITLE
support CP for KimiK3

### PR DESCRIPTION
**Pull request by acisseJZhong**
## Summary

Enable context parallelism for Kimi K3's hybrid KDA/MLA decoder and
multimodal input path.

- KDA layers use Attention Gym's native context-parallel implementation. Each
  rank keeps its local sequence fragments while exchanging the short-convolution
  history and composing the recurrent state from preceding fragments.
- MLA layers can use all-gather K/V CP with contiguous or headtail partitions,
  or Ulysses CP with contiguous partitions.
- Multimodal preprocessing shards token-aligned vision-bank indices with the
  token IDs, allowing each rank to map replicated vision-encoder outputs to its
  local placeholders.

KDA and MLA reuse the same TorchTitan CP mesh. The all-gather configuration
supports contiguous and headtail partitions across the hybrid model; the
Ulysses configuration requires contiguous partitioning.

## Implementation

The integration follows Attention Gym's [delta-rule context-parallel example](https://github.com/meta-pytorch/attention-gym/blob/main/examples/linear/delta_rule_context_parallel.py).

The KDA path separates its local stages so the CP adapter can replace only the
distributed operations while sharing projection, normalization, gating, and
output logic with the non-CP path. Attention Gym provides:

- `context_parallel_conv_history` for convolution halo exchange; and
- `context_parallel_kda` for recurrent-state summary composition and the local
  KDA computation.

TorchTitan builds an Attention Gym routing plan that matches its contiguous or
headtail token assignment and passes the per-batch routing through every KDA
layer.

## Limitations

- KDA CP currently supports contiguous and headtail partitions; PTRR is not
  supported.
- Ulysses MLA requires contiguous partitioning; it does not support headtail.
- Current TorchTitan partitioning requires equal-sized fragments, where `T` is
  the token count and `C` is the CP degree:
  - contiguous: `T % C == 0`
  - headtail: `T % (2 * C) == 0`

  This is a limitation of the current PyTorch/TorchTitan equal-sized sharder,
  not Attention Gym; Attention Gym routing accepts arbitrary fragment ranges.
- Kimi K3 sample packing is not yet supported. Routing currently models the
  batch as one global sequence.

## Tests

Tested on NVIDIA H100 GPUs with Attention Gym main at commit `7418529`.

Targeted unit tests:

```text
pytest -q tests/unit_tests/gpu/test_kda_attention.py \
  tests/unit_tests/gpu/test_kimi_k3.py
# 13 passed

pytest -q tests/unit_tests/cpu/test_integration_test_definitions.py
# 12 passed
```

Two-GPU integration tests (10 training steps each):

```text
python -m tests.integration_tests.run_tests <output-dir> \
  --test_suite b200 \
  --test_name kimi_k3_mm_allgather_kv_cp \
  --ngpu 2 --no-parallel
# rc=0

python -m tests.integration_tests.run_tests <output-dir> \
  --test_suite b200 \
  --test_name kimi_k3_mm_ulysses_cp \
  --ngpu 2 --no-parallel
# rc=0
```

Deterministic BF16 runs used the same seed checkpoint (`seed=42`), 256 tokens
per step, and 100 training steps. CP=1 used one GPU; CP=2 used two GPUs with
Attention Gym CP for KDA and either headtail all-gather K/V or contiguous
Ulysses for MLA. Values below are measured at steps 1, 10, and 100. Each diff
is the absolute percentage difference from CP=1.

| Step | CP=1 loss | All-gather loss | Diff | Ulysses loss | Diff |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | `12.624805` | `12.616027` | `0.070%` | `12.627501` | `0.021%` |
| 10 | `4.031361` | `3.878002` | `3.804%` | `4.639540` | `15.086%` |
| 100 | `2.794998` | `2.930616` | `4.852%` | `2.846095` | `1.828%` |

| Step | CP=1 grad norm | All-gather grad norm | Diff | Ulysses grad norm | Diff |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | `25.625` | `25.625` | `0.000%` | `25.75` | `0.488%` |
| 10 | `6.84375` | `6.96875` | `1.826%` | `11.625` | `69.863%` |
| 100 | `5.75` | `5.3125` | `7.609%` | `5.8125` | `1.087%` |

### CODE REVIEW

File: .ci/docker/requirements.txt [hunk 1]
```diff
@@ -8,4 +8,4 @@ safetensors
 einops
 pillow
 spmd_types==0.2.5
-attn-gym[linear]==0.0.5
+attn-gym[linear] @ git+https://github.com/meta-pytorch/attention-gym.git@main
```
Comment (drisspg) [inline .ci/docker/requirements.txt:13]: lets change before land; we are also on 0.0.8 ( that should be checked in if it hasnt its probs a bug; ) but the  0.8 didnt have the cp stuff finalzied we can cut a new one for this

File: torchtitan/models/kimi_k3/model.py [hunk 2]
```diff
@@ -315,6 +360,95 @@ def __init__(self, config: Config):
...
+            if len(self.kda_conv_kernel_sizes) != 1:
+                raise ValueError(
+                    "Kimi K3 context parallelism requires every KDA layer "
+                    "to use the same convolution kernel size."
+                )
+            cp_mesh = parallel_dims.get_mesh("cp")
+            tokens = batch.get("input")
+            if not isinstance(tokens, torch.Tensor):
+                raise ValueError("Kimi K3 context parallelism requires token input.")
+            batch["kda_cp_routing"] = build_kda_context_parallel_routing(
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k3/model.py]: Code looks a bit hard to follow. E.g. are you making `kda_cp_routing` field sharded in this function, and would the `cp_shard` in the function below shard it again? https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/context_parallel/api.py#L109

Following my comment https://github.com/pytorch/torchtitan/pull/4450#discussion_r3942280306, I think we should abstract the CP sharding method to associate it with each Attention / CP backend, and here in `model.preprocess_inputs` we should only call succinct API interface.

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/models/kimi_k3/model.py (hunk 2 above)
Comment (acisseJZhong) [reply torchtitan/models/kimi_k3/model.py]: no cp_shard will only shard tensor inputs. So this routing object will not be sharded again. https://github.com/pytorch/torchtitan/blob/5bc288c290ba97933391dc93b19084258f0bb830/torchtitan/distributed/context_parallel/api.py#L98

Yes maybe we can customize cp_shard for different backends, and put `build_kda_context_parallel_routing` into cp_shard for kda cp. cc @fegin 

the rest of this function is pretty standard, following muse glimmer's preprocess input.

File: tests/unit_tests/gpu/test_kimi_k3.py [hunk 3]
```diff
@@ -115,8 +245,8 @@ def test_flex_attention_mask(self):
 
     @unittest.skipIf(
         not torch.cuda.is_available()
-        or torch.cuda.get_device_capability() not in {(10, 0), (10, 3)},
-        "Attention Gym KDA requires CUDA capability 10.0 or 10.3.",
+        or torch.cuda.get_device_capability() not in {(9, 0), (10, 0), (10, 3)},
```
Comment (drisspg) [inline tests/unit_tests/gpu/test_kimi_k3.py:260]: we do support ampere+ CP might need hopper + though

File: torchtitan/models/kimi_k3/model.py [hunk 4]
```diff
@@ -278,6 +297,32 @@ def update_from_config(self, *, config, **kwargs) -> None:
             if isinstance(dataset, MMSamplePackingConfig):
                 raise ValueError("Kimi K3 does not yet support sample packing.")
             Decoder.Config.update_from_config(self, config=config, **kwargs)
+            parallelism = config.parallelism
+            if parallelism.context_parallel_degree > 1 and (
+                parallelism.context_parallel_load_balancer not in (None, "headtail")
```
Comment (drisspg) [inline torchtitan/models/kimi_k3/model.py]: Is there a nice way to register some other form of balancer? Or at least a plan to open this up

File: torchtitan/models/kimi_k3/cp_kda.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+from torchtitan.transforms.base import retype_node
+
+from .kda import InnerKDA, KDA
+
+
+def partition_fragments(
+    num_tokens: int,
+    world_size: int,
+    load_balancer: str | None,
+) -> list[list[tuple[int, int]]]:
```
Comment (drisspg) [inline torchtitan/models/kimi_k3/cp_kda.py:31]: Yeah, I do think we should think about how we want people to extend this. Could be a follow-up

File: torchtitan/models/kimi_k3/model.py (hunk 4 above)
Comment (fegin) [reply torchtitan/models/kimi_k3/model.py]: Do you mean that users can add their own load balancer? The corresponding CP kernel needs to know the load-balancing algorithm. So they need to swap together, the interface is not clean yet. After the stack of https://github.com/pytorch/torchtitan/pull/4459, it may be a step closer, but I haven't fully thought about that.

What's the load balancing in your mind? cc., @jinsooihm

Comment (jinsooihm) [reply torchtitan/models/kimi_k3/model.py]: Maybe each kernel can define their load balancing function, which can be passed in `prepare_context_parallel_input`? But there could be some problems if a model uses two different kernels and their load balancers are not compatible.

**Review by fegin — COMMENTED** [review]
As discussed offline, the Ulysses introduced in https://github.com/pytorch/torchtitan/pull/4450 is not the best one for MLA attention. You may want to subclass the UlyssesCPFlexInnerAttention to provide a new one.

File: torchtitan/models/kimi_k3/model.py (hunk 4 above)
Comment (drisspg) [reply torchtitan/models/kimi_k3/model.py]: so at least at the kda level we just need the dataloader/batcher to describe the `fragment allocation per rank` there are alot of nouns and verbs I have a detailed walkthrough (quite long) https://github.com/meta-pytorch/attention-gym/blob/81c186d900658c9da2597acaa1042f1b8721ffe8/attn_gym/linear/context_parallel.py#L7

my original claim is that depending on your actual distribution of packed documents you might want to do different load balancing. This is not really applicable today with just the Dense path supported so its moot until then but food for thought

File: torchtitan/models/kimi_k3/cp_kda.py (hunk 5 above)
Comment (acisseJZhong) [reply torchtitan/models/kimi_k3/cp_kda.py:31]: added a todo

**Comment by acisseJZhong** [comment]
Replacement upstream-branch PR: #4639. This preserves the same stack base while allowing GitHub native PR stacks.
